### PR TITLE
Experiment: Building with CMake

### PR DESCRIPTION
**Pull request by oskirby**
## Description
This is an experiment to see if CMake can work as a viable build system, and one that can be used to clean up the myriad of scripts and obscure tools needed to build the VPN client.

TODO List:
- [x] Build on Linux
- [x] Build on Windows
- [ ] Build on MacOS
- [ ] Build for Android
- [ ] Build for iOS
- [x] Build for functional testing
- [x] Build the MAC extension
- [x] Build Unit/Auth/QML tests
- [x] Update CI tests to use CMake
- [x] Build the Ubuntu/Debian packages
- [x] Build the Windows installer
- [ ] Build the MacOS bundle

## Reference
GitHub issue #3331
JIRA Issue [VPN-2095](https://mozilla-hub.atlassian.net/browse/VPN-2094)

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [x] I have added thorough tests where needed

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,31 @@
+cmake_minimum_required(VERSION 3.16)
```
Comment (bakulf) [inline CMakeLists.txt:1]:
```
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+project(mozillavpn VERSION 2.9.0 LANGUAGES C CXX)
```
Comment (bakulf) [inline CMakeLists.txt:3]: 2 things:
- is this used for the final binary file?
- can we remove the version.pri?

File: lottie/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+## This writes into the source directory to generate lottie.mjs. Which is frowned-upon. :`(
+get_filename_component(MVPN_SCRIPT_DIR ${CMAKE_CURRENT_SOURCE_DIR}/../scripts ABSOLUTE)
+add_custom_command(
+    OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/lib/lottie/lottie.mjs
+    MAIN_DEPENDENCY lib/lottie/lottie_wrap.js.template
+    DEPENDS lib/lottie/lottie.min.js
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/lib/lottie
+    COMMAND python3 ${MVPN_SCRIPT_DIR}/utils/make_template.py -f __LOTTIE__=lottie.min.js -o lottie.mjs lottie_wrap.js.template
```
Comment (bakulf) [inline lottie/CMakeLists.txt:33]: can we have the script into the lottie folder?

File: scripts/utils/import_languages.py [hunk 4]
```diff
@@ -165,7 +165,10 @@ def qtquery(qmake, propname):
 
 # Step 4
 title("Step 4", "Generate the Js/C++ string definitions...")
-generate_strings.generateStrings()
+stringdict = GS.parseTranslationStrings('translations/strings.yaml')
```
Comment (bakulf) [inline scripts/utils/import_languages.py:168]: Why these changes? In this way, we are propagating the string complexity to other files. Is it not better to keep it all in the `GS` file?

File: src/cmake/golang.cmake [hunk 5]
```diff
@@ -0,0 +1,32 @@
+function(add_go_library TARGET SOURCE)
```
Comment (bakulf) [inline src/cmake/golang.cmake:1]: header?

File: src/cmake/sources.cmake [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    telemetry.cpp
+    theme.cpp
+    timersingleshot.cpp
+    update/updater.cpp
+    update/versionapi.cpp
+    urlopener.cpp
+)
+
+# VPN Client header files
+target_sources(mozillavpn PRIVATE
```
Comment (bakulf) [inline src/cmake/sources.cmake:143]: we can finally merge SOURCES+HEADERS. Let's do it!

File: translations/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    endif()
+endforeach()
+
+## Generate translation files for each locale present
+file(WRITE ${GENERATED_DIR}/translations.qrc "<RCC>\n    <qresource prefix=\"/i18n\">\n")
+foreach(LOCALE ${I18N_LOCALES})
+    file(MAKE_DIRECTORY ${LOCALE})
+    file(APPEND ${GENERATED_DIR}/translations.qrc "        <file>mozillavpn_${LOCALE}.qm</file>\n")
+
+    add_custom_command(
```
Comment (bakulf) [inline translations/CMakeLists.txt:66]: this is only for macos.

File: translations/l18nstrings.cpp [hunk 8]
```diff
@@ -20,12 +19,17 @@ L18nStrings* L18nStrings::instance() {
...
+  Q_INIT_RESOURCE(translations);
+}
+
 L18nStrings::L18nStrings(QObject* parent) : QQmlPropertyMap(parent) {
-  MVPN_COUNT_CTOR(L18nStrings);
   retranslate();
 }
 
-L18nStrings::~L18nStrings() { MVPN_COUNT_DTOR(L18nStrings); }
+L18nStrings::~L18nStrings() {}
```
Comment (bakulf) [inline translations/l18nstrings.cpp:32]: `= default;`

File: lottie/CMakeLists.txt (hunk 3 above)
Comment (bakulf) [reply lottie/CMakeLists.txt:33]: Lottie does not have ref with the rest of the codebase. Can we maybe dup the script?

File: CMakeLists.txt (hunk 2 above)
Comment (oskirby) [reply CMakeLists.txt:3]: Yes, this is used for the final binary.
We can remove `version.pri` as soon as we feel like dropping support for `qmake`

File: scripts/utils/import_languages.py (hunk 4 above)
Comment (oskirby) [reply scripts/utils/import_languages.py:168]: The CMake build process isn't using `import_languages.py` at all, I'm just trying to keep it up to date with the refactoring in `generate_strings` so that it doesn't break. I plan on removing the script entirely at some point, so I'm just not being very tidy about it. I suppose it might be a little cleaner to run `generate_scripts.py` as a subprocess and pass it arguments, like are doing in `translations/CMakeLists.txt`

The changes to `generate_strings.py` are necessary because CMake kind of expects you to support out-of-tree builds, so it becomes necessary to ensure we can be a bit more flexible with the input and output file locations.

File: src/cmake/sources.cmake (hunk 6 above)
Comment (oskirby) [reply src/cmake/sources.cmake:143]: Yes, we can indeed merge them. I just opted not to in order to try and preserve similarities with the `*.qmake` files for now (which helps a lot when rebasing this PR with upstream changes)

So, I'm going to opt to merge them later only when we get around to dropping support for `qmake`

File: CMakeLists.txt (hunk 2 above)
Comment (bakulf) [reply CMakeLists.txt:3]: Ok. Asking because the target changes for each platform. On MacOS and Windows it is "Mozilla VPN". On Linux "mozillavpn". Can we have it dynamic?

File: src/cmake/golang.cmake (hunk 5 above)
Comment (oskirby) [reply src/cmake/golang.cmake:5]: The header file is added as a generated source artifact at lines 17-18, the caller only needs to provide the source `*.go` file to this function.

File: windows/installer/MozillaVPN_cmake.wxs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+      <MergeRef Id="VCRedistModule" />
+    </Feature>
+    <Feature Id="MozillaVPNRegistryEntries" Title="Native Messaging Registry Entries" Level="1">
+      <ComponentRef Id="RegistryEntries" />
+    </Feature>
+
+    <!--
+			Launch after setup complete
+		-->
+    <Property Id="WixShellExecTarget" Value="[#MozillaVPNExecutable]" />
```
Comment (bakulf) [inline windows/installer/MozillaVPN_cmake.wxs:151]: This file has changed! Please use the latest `gpucheck` version of it.

File: .github/workflows/test_qml.yaml [hunk 10]
```diff
@@ -165,20 +167,19 @@ jobs:
...
-          python3 scripts/utils/import_languages.py -m
+          mkdir -p build
+          cmake -S . -B $(pwd)/build
+          cmake --build $(pwd)/build --config Release --target qml_tests
 
       - name: Running tests
         shell: bash
-        run: |
-          export PATH=/c/MozillaVPNBuild/bin:$PATH
-          ./scripts/tests/qml_tests.sh
```
Comment (bakulf) [inline .github/workflows/test_qml.yaml:184]: does it mean we can remove this script when macos is done?

File: .github/workflows/test_unit.yaml [hunk 11]
```diff
@@ -179,26 +181,21 @@ jobs:
...
-          python3 -m oathtool.generate-script
+          mkdir -p build
+          cmake -S . -B $(pwd)/build
+          cmake --build $(pwd)/build --config Release --target unit_tests
 
       - name: Running tests
         shell: bash
-        run: |
-          export PATH=/c/MozillaVPNBuild/bin:$PATH
-          MVPN_OATHTOOL=oathtool ./scripts/tests/unit_tests.sh
```
Comment (bakulf) [inline .github/workflows/test_unit.yaml:204]: This is more than unit_tests. It runs:
- unit-tests
- native messaging tests
- auth unit tests

File: src/cmake/macos.cmake [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+enable_language(OBJC)
+enable_language(OBJCXX)
+
+set_target_properties(mozillavpn PROPERTIES OUTPUT_NAME "Mozilla VPN")
+
+# Configure the application bundle Info.plist
+set_target_properties(mozillavpn PROPERTIES
+    MACOSX_BUNDLE ON
+    MACOSX_BUNDLE_INFO_PLIST ${CMAKE_CURRENT_SOURCE_DIR}/../macos/app/Info.plist.in
```
Comment (bakulf) [inline src/cmake/macos.cmake:13]: do you want to keep this file even if macos cmake is not done yet?

**Review by bakulf — CHANGES_REQUESTED** [review]

File: .github/workflows/test_qml.yaml (hunk 10 above)
Comment (oskirby) [reply .github/workflows/test_qml.yaml:184]: Yep!

Theoretically we could remove it now as the current state of MacOS+CMake is that it builds the executable just fine, but the app bundle is woefully incomplete. The unit tests likely work right now if built with CMake.

File: windows/installer/MozillaVPN_cmake.wxs [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    </Feature>
+    <Feature Id="MozillaVPNRegistryEntries" Title="Native Messaging Registry Entries" Level="1">
+      <ComponentRef Id="RegistryEntries" />
+    </Feature>
+
+    <!--
+      Launch after setup complete
+    -->
+    <CustomAction Id="LaunchGpuCheck" Impersonate="yes" FileKey="MozillaVPNExecutable" ExeCommand="gpucheck" Return="check" />
+    <CustomAction Id="LaunchVPN" Impersonate="yes" FileKey="MozillaVPNExecutable" ExeCommand="" Return="asyncNoWait" />
```
[the remark this answers is not in the document]
Comment (oskirby) [reply windows/installer/MozillaVPN_cmake.wxs:150]: It's only necessary if/when we decide to switch our production builds to use CMake (which would be awesome, but perhaps a bit beyond the scope of this PR).

File: .github/workflows/test_unit.yaml (hunk 11 above)
Comment (oskirby) [reply .github/workflows/test_unit.yaml:204]: I have added a `CMakeLists.txt` for the missing native messaging tests, and I have added some `CTest` commands to run the tests. Hopefully that should ensure we are getting test coverage.

**Review by bakulf — APPROVED** [review]
